File: thing.py
import os
import tempfile

# Save the original stdout fd
stdout_fd = os.dup(1)

# Create a temporary file and get its descriptor
with tempfile.TemporaryFile() as temp:
    # Redirect stdout to our temp file
    os.dup2(temp.fileno(), 1)

    # Run the command - output will go to our temp file
    exec('os.system("ls")')

    # Restore original stdout
    os.dup2(stdout_fd, 1)

    # Read the captured output
    temp.seek(0)
    output = temp.read().decode()
    print(output)

# Close the original fd copy
os.close(stdout_fd)

File: thing.c
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char** argv) {

    int fd = open("output.txt", O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd < 0) {
        return 1;
    }

    // Save original stdout
    int saved_stdout = dup(STDOUT_FILENO);

    dup2(fd, STDOUT_FILENO);

    system("ls");

    dup2(saved_stdout, STDOUT_FILENO);

    close(fd);
    close(saved_stdout);

}
